cps/elf.c

```diff
@@ -0,0 +1,406 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <elf.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <string.h>
+#include <fcntl.h>
+
+#include "gatekeeper_cps.h"
+#include "elf.h"
+
+#if __WORDSIZE == 32
+#define PERBIT(x) x##32
+#define ElfPERBIT(x) Elf32_##x
+#else
+#define PERBIT(x) x##64
+#define ElfPERBIT(x) Elf64_##x
+#endif
+
+struct elf_file {
+	/* Path to ELF file. */
+	char *pathname;
+
+	/* Whether the endianness needs to be converted. */
+	int  conv;
+
+	/* File contents. */
+	void *data;
+
+	/* Length of file contents. */
+	unsigned long len;
+};
+
+struct string_table
+{
+	/* Number of strings in the table. */
+	unsigned int cnt;
+
+	/* Maximum number of strings allowed in table. */
+	unsigned int max;
+
+	/* Start of strings. */
+	const char *str[0];
+};
+
+/*
+ * Get CPU endianness.
+ *
+ * 0 = unknown
+ * 1 = ELFDATA2LSB = little
+ * 2 = ELFDATA2MSB = big
+ */
+static int __attribute__((pure))
+native_endianness(void)
+{
+	/*
+	 * Encoding the endianness enums in a string
+	 * and then reading that string as a 32-bit int,
+	 * returns the correct endianness automagically.
+	 */
+	return (char)*((const uint32_t *)("\1\0\0\2"));
+}
+
+/* Change endianness of x if @conv is true. */
+#define END(x, conv)							\
+({									\
+	typeof(x) __x;							\
+	if (conv) __swap_bytes(&(x), &(__x), sizeof(__x));		\
+	else __x = (x);							\
+	__x;								\
+})
+
+#define strstarts(a, start) (strncmp((a), (start), strlen(start)) == 0)
+
+static inline void
+__swap_bytes(const void *src, void *dest, unsigned int size)
+{
+	unsigned int i;
+	for (i = 0; i < size; i++)
+		((unsigned char *)dest)[i] =
+			((const unsigned char *)src)[size - i - 1];
+}
+
+/* Get the section specified in @secname from the ELF file. */
+static void *
+PERBIT(get_section)(struct elf_file *module, const char *secname,
+	unsigned long *secsize)
+{
+	void *data = module->data;
+	unsigned long len = module->len;
+	int conv = module->conv;
+
+	unsigned int i;
+	const char *secstrings;
+
+	ElfPERBIT(Ehdr) *hdr = data;
+	ElfPERBIT(Shdr) *sechdrs;
+	ElfPERBIT(Off) e_shoff;
+	ElfPERBIT(Half) e_shnum, e_shstrndx;
+
+	*secsize = 0;
+
+	/* Check whether there's enough room for ELF header. */
+	if (len <= 0 || len < sizeof(*hdr))
```

Comment:
I think we can remove the condition ``len <= 0``, as the condition ``len < sizeof(*hdr)`` contains that scenario.